File: PageObjectJava/src/test/java/PageObjects/AutomationPracticeStore/Objects/StoreItemThumbnail.java
package PageObjects.AutomationPracticeStore.Objects;

import Base.TestConfiguration;
import PageObjects.AutomationPracticeStore.ErrorMessageException;
import PageObjects.AutomationPracticeStore.ProductPage;
import Support.Support;
import com.google.common.base.Strings;
import org.openqa.selenium.By;
import org.openqa.selenium.NoSuchElementException;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.support.FindBy;
import org.openqa.selenium.support.How;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

public class StoreItemThumbnail {

    //Regular web elements found in the items across all screens

    @FindBy(how = How.CSS, using = "a.product_img_link img")
    private WebElement itemImage;

    @FindBy(how = How.TAG_NAME, using = "h5")
    private WebElement itemName;

    @FindBy(how = How.CSS, using = "div.right-block .price")
    private WebElement itemPrice;

    @FindBy(how = How.CSS, using = "div.right-block .price-percent-reduction")
    private List<WebElement> itemDiscount;

    @FindBy(how = How.CSS, using = "div.right-block .old-price")
    private List<WebElement> itemPriceBeforeDiscount;

    /* The following web elements may already exist in the DOM but are not clickable.
       Should be accessed only after hovering the mouse over the item itself. */

    @FindBy(how = How.CSS, using = "div.left-block .price")
    private WebElement itemContextPrice;

    @FindBy(how = How.CSS, using = "div.left-block .price-percent-reduction")
    private List<WebElement> itemContextDiscount;

    @FindBy(how = How.CSS, using = "div.left-block .old-price")
    private List<WebElement> itemContextPriceBeforeDiscount;

    @FindBy(how = How.CSS, using = "div.right-block a.ajax_add_to_cart_button")
    private WebElement itemContextAddToCartButton;

    @FindBy(how = How.CSS, using = "div.right-block a.lnk_view")
    private WebElement itemContextMoreButton;

    @FindBy(how = How.CLASS_NAME, using = "quick-view")
    private WebElement itemQuickViewButton;


    /* TODO: Create the proper exception in the methods handling all the following locators
       Web Elements with extra functionality that only appear outside of the Home screen */

    @FindBy(how = How.CLASS_NAME, using = "color-pick")
    private List<WebElement> itemExtraColorPicking;

    @FindBy(how = How.CLASS_NAME, using = "availability")
    private List<WebElement> itemStockStatus;

    @FindBy(how = How.CLASS_NAME, using = "addToWishlist")
    private List<WebElement> itemAddToWishListTab;    //Needs to be hover over the item first

    @FindBy(how = How.CLASS_NAME, using = "add_to_compare")
    private WebElement itemAddToCompareTab;     //Needs to be hover over the item first

    private final static String IN_STOCK_MESSAGE = "In stock";
    private final static String ADDED_TO_WISH_LIST = "Added to your wishlist.";

    private final static By ERROR_MESSAGE_BOX_LOCATOR = By.className("fancybox-error");

    public String getName(){
        return this.itemName.getText();
    }

    public double getPrice(){
        return Double.parseDouble(this.itemPrice.getText().substring(1));
    }

    public String getImageUrl(){
        return this.itemImage.getAttribute("src");
    }

    public int getDiscountPercentage(){
        if (this.itemDiscount.isEmpty())
            return 0;
        String txt = this.itemDiscount.get(0).getText().split("%")[0].substring(1);
        return Integer.parseInt(txt);
    }

    public double getPriceBeforeDiscount(){
        if(this.itemPriceBeforeDiscount.isEmpty())
            return 0;
        return Double.parseDouble(this.itemPriceBeforeDiscount.get(0).getText().substring(1));
    }

    public double getContextPrice(){
        Support.hoverOverWebElement(this.itemImage);
        return Double.parseDouble(this.itemContextPrice.getText().substring(1));
    }

    public int getContextDiscountPercentage(){
        Support.hoverOverWebElement(this.itemImage);
        if(this.itemContextDiscount.isEmpty())
            return 0;
        return Integer.parseInt(this.itemContextDiscount.get(0).getText().split("%")[0].substring(1));
    }

    public double getContextPriceBeforeDiscount(){
        Support.hoverOverWebElement(this.itemImage);
        if(this.itemContextPriceBeforeDiscount.isEmpty())
            return 0;
        return Double.parseDouble(this.itemContextPriceBeforeDiscount.get(0).getText().substring(1));
    }

    public ProductInCartPopUp addToCart(){
        Support.hoverOverWebElement(this.itemImage);
        this.itemContextAddToCartButton.click();
        return new ProductInCartPopUp();
    }

    public ProductPage moreInformation(){
        Support.hoverOverWebElement(this.itemImage);
        this.itemContextMoreButton.click();
        return new ProductPage();
    }

    public QuickView getQuickView(){
        Support.hoverOverWebElement(this.itemImage);
        this.itemQuickViewButton.click();
        return new QuickView();
    }

    // The following methods should be used only if the item is being handled by a CatalogPage instance

    public boolean isInStock(){
            if(this.itemStockStatus.isEmpty())
                throw new NoSuchElementException("The label for stock availability was not found. Make sure this object " +
                        "is being displayed in the catalog page");
        return this.itemStockStatus.get(0).getText().trim().equals(IN_STOCK_MESSAGE);
    }

    public List<String> getAvailableColors(){
        List<String> availableColors = new ArrayList<>();
        String colorHexCode = null;
        for(WebElement colorWe : itemExtraColorPicking){
            try{
                colorHexCode = colorWe.getAttribute("style").split("#")[1];
            }catch(IndexOutOfBoundsException ex){
                Support.LOGGER.error(String.format("The item named '%s' has one color missing", this.getName()));
            }
            if(!Strings.isNullOrEmpty(colorHexCode))
                availableColors.add(colorHexCode);
        }
        return availableColors;
    }

    public boolean isProductInColor(String colorHexCode){
        return this.getAvailableColors().stream().anyMatch(color -> color.equalsIgnoreCase(colorHexCode));
    }

    public ProductPage selectColor(String colorHexCode){
        if(this.getAvailableColors().isEmpty())
            throw new NoSuchElementException(String.format("The item '%s' has not colors to display. " +
                    "Make sure this object is being displayed in the Catalog page or the color section exists",
                    this.getName()));
        WebElement found = this.itemExtraColorPicking.stream().filter(colorWe -> colorWe.getAttribute("style").equalsIgnoreCase(colorHexCode))
                .findFirst()
                .orElseThrow(() -> new NoSuchElementException(String.format("The item named '%s' is not available in color '%s'", this.getName(), colorHexCode)));
        found.click();
        return new ProductPage();
    }

    public StoreItemThumbnail addToWishList(){
        Support.hoverOverWebElement(this.itemImage);
        if(this.itemAddToWishListTab.isEmpty()){
            String errorText = String.format("The tab for adding to the wish list for item '%s' was not found. " +
                    "Make sure this object is being displayed in the Catalog page", this.getName());
            throw new NoSuchElementException(errorText);
        }

        this.itemAddToWishListTab.get(0).click();

        List<WebElement> msg = TestConfiguration.getDriver().findElements(ERROR_MESSAGE_BOX_LOCATOR);
        if(msg.isEmpty())
            throw new NoSuchElementException("The expected pop up message did not appear");
        if(!msg.get(0).getText().equals(ADDED_TO_WISH_LIST))
            throw new ErrorMessageException(msg.get(0).getText());
        msg.get(0).findElement(By.xpath("../../../a[@title='Close']")).click(); //For closing the popup message that should always appear
        return this;
    }

    public boolean isAddedToWishList(){
        Support.hoverOverWebElement(this.itemImage);
        if(this.itemAddToWishListTab.isEmpty()){
            String errorText = String.format("The tab for adding to the wish list for item '%s' was not found. " +
                    "Make sure this object is being displayed in the Catalog page", this.getName());
            throw new NoSuchElementException(errorText);
        }
        String attr = this.itemAddToWishListTab.get(0).getAttribute("class");
        return attr.endsWith("checked");
    }

    public StoreItemThumbnail addToCompare(boolean addToCompare){
        Support.hoverOverWebElement(this.itemImage);

        if(addToCompare && !this.isAddedToCompare())
            this.itemAddToCompareTab.click();
        if(!addToCompare && this.isAddedToCompare())
            this.itemAddToCompareTab.click();
        return this;
    }

    public boolean isAddedToCompare(){
        Support.hoverOverWebElement(this.itemImage);
        if(Objects.isNull(this.itemAddToCompareTab)){
            String errorText = String.format("The tab for adding for comparison for item '%s' was not found. " +
                    "Make sure this object is being displayed in the Catalog page", this.getName());
            throw new NoSuchElementException(errorText);
        }
        String classAttribute = this.itemAddToCompareTab.getAttribute("class");
        return classAttribute.endsWith("checked");
    }
}

File: PageObjectDotNet/AutomationPractice/PageObjects/AutomationPracticeStore/Objects/StoreItemThumbnail.cs
using OpenQA.Selenium;
using OpenQA.Selenium.Support.PageObjects;
using SeleniumFramework.Base;
using System;
using System.Collections.Generic;
using System.Linq;

namespace AutomationPractice.PageObjects.AutomationPracticeStore.Objects
{
    public class StoreItemThumbnail
    {

        //Regular web elements found in the items across all screens

        [FindsBy(How = How.CssSelector, Using = "a.product_img_link img")]
        private IWebElement itemImage;

        [FindsBy(How = How.TagName, Using = "h5")]
        private IWebElement itemName;

        [FindsBy(How = How.CssSelector, Using = "div.right-block .price")]
        private IWebElement itemPrice;

        [FindsBy(How = How.CssSelector, Using = "div.right-block .price-percent-reduction")]
        private IList<IWebElement> itemDiscount;

        [FindsBy(How = How.CssSelector, Using = "div.right-block .old-price")]
        private IList<IWebElement> itemPriceBeforeDiscount;

        /* The following web elements may already exist in the DOM but are not clickable.
           Should be accessed only after hovering the mouse over the item itself. */

        [FindsBy(How = How.CssSelector, Using = "div.left-block .price")]
        private IWebElement itemContextPrice;

        [FindsBy(How = How.CssSelector, Using = "div.left-block .price-percent-reduction")]
        private IList<IWebElement> itemContextDiscount;

        [FindsBy(How = How.CssSelector, Using = "div.left-block .old-price")]
        private IList<IWebElement> itemContextPriceBeforeDiscount;

        [FindsBy(How = How.CssSelector, Using = "div.right-block a.ajax_add_to_cart_button")]
        private IWebElement itemContextAddToCartButton;

        [FindsBy(How = How.CssSelector, Using = "div.right-block a.lnk_view")]
        private IWebElement itemContextMoreButton;

        [FindsBy(How = How.ClassName, Using = "quick-view")]
        private IWebElement itemQuickViewButton;


        /* TODO: Create the proper exception in the methods handling all the following locators
           Web Elements with extra functionality that only appear outside of the Home screen */

        [FindsBy(How = How.ClassName, Using = "color-pick")]
        private IList<IWebElement> itemExtraColorPicking;

        [FindsBy(How = How.ClassName, Using = "availability")]
        private IList<IWebElement> itemStockStatus;

        [FindsBy(How = How.ClassName, Using = "addToWishlist")]
        private IList<IWebElement> itemAddToWishListTab;    //Needs to be hover over the item first

        [FindsBy(How = How.ClassName, Using = "add_to_compare")]
        private IWebElement itemAddToCompareTab;     //Needs to be hover over the item first

        private readonly static string IN_STOCK_MESSAGE = "In stock";
        private readonly static string ADDED_TO_WISH_LIST = "Added to your wishlist.";

        private readonly static By ERROR_MESSAGE_BOX_LOCATOR = By.ClassName("fancybox-error");



        public string GetName() { return itemName.Text; }

        public float GetPrice() { return float.Parse(itemPrice.Text.Substring(1)); }

        public string GetImageURL() 
        {
            return itemImage.GetAttribute("src");
        }

        public int GetDiscountPercentage() 
        {
            return itemDiscount.Count == 0 ? 0 : int.Parse(itemDiscount[0].Text.Split('%')[0].Substring(1));
        }

        public float GetPriceBeforeDiscount() 
        {
            return itemPriceBeforeDiscount.Count == 0 ? 0 : float.Parse(itemPriceBeforeDiscount[0].Text.Substring(1));
        }

        public float GetContextPrice() 
        {
            Support.HoverOverWebElement(TestConfiguration.GetDriverInstance(), itemImage);
            return float.Parse(itemContextPrice.Text.Substring(1));
        }

        public int GetContextDiscountPercentage() 
        {
            Support.HoverOverWebElement(TestConfiguration.GetDriverInstance(), itemImage);
            return itemContextDiscount.Count == 0 ? 0 : int.Parse(itemContextDiscount[0].Text.Split('%')[0].Substring(1));
        }

        public float GetContextPriceBeforeDiscount() 
        {
            Support.HoverOverWebElement(TestConfiguration.GetDriverInstance(), itemImage);
            return itemContextPriceBeforeDiscount.Count == 0 ? 0 : float.Parse(itemContextPriceBeforeDiscount[0].Text.Substring(1));
        }


        public ProductInCartPopUp AddToCart() 
        {
            Support.HoverOverWebElement(TestConfiguration.GetDriverInstance(), itemImage);
            itemContextAddToCartButton.Click();
            return new ProductInCartPopUp();
        }

        public ProductPage MoreInformation() 
        {
            Support.HoverOverWebElement(TestConfiguration.GetDriverInstance(), itemImage);
            itemContextMoreButton.Click();
            return new ProductPage();
        }

        public QuickView GetQuickView() 
        {
            Support.HoverOverWebElement(TestConfiguration.GetDriverInstance(), itemImage);
            itemQuickViewButton.Click();
            return new QuickView();
        }


        // The following methods should be used only if the item is being handled by a CatalogPage instance

        public bool IsInStock() 
        {
            if (itemStockStatus.Count == 0)
                throw new NoSuchElementException("The label for stock availability was not found. Make sure this object is being displayed in the Catalog page");
            return itemStockStatus[0].Text.Trim().Equals(IN_STOCK_MESSAGE);
        }

        public IReadOnlyList<string> GetAvailableColors() 
        {
            var availableColors = new List<string>();
            string colorHexCode = string.Empty;
            itemExtraColorPicking.ToList().ForEach(colorWe => {
                try
                {
                    colorHexCode = colorWe.GetAttribute("style").Split('#')[1];
                }
                catch (IndexOutOfRangeException)
                {
                    Console.Error.WriteLine($"The item named '{ GetName() }' has one color missing.");
                }
                if (!string.IsNullOrWhiteSpace(colorHexCode))
                    availableColors.Add(colorHexCode);
            });
            return availableColors.AsReadOnly();
        }

        public bool IsProductInColor(string colorHexCode) 
        {
            return GetAvailableColors().Any(color => string.Equals(color, colorHexCode, StringComparison.CurrentCultureIgnoreCase));
        }

        public ProductPage SelectColor(string colorHexCode) 
        {
            if (GetAvailableColors().Count == 0)
                throw new NoSuchElementException($"The item '{ GetName() }' has not colors to display. " +
                    "Make sure this object is being displayed in the Catalog page or the color section exists");

            var found = itemExtraColorPicking.Where(colorWe => string.Equals(colorWe.GetAttribute("style"), colorHexCode, StringComparison.CurrentCultureIgnoreCase)).FirstOrDefault() 
                ?? throw new NoSuchElementException($"The item named '{ GetName() }' is not available in color '{ colorHexCode }'");

            found.Click();
            return new ProductPage();
        }

        public StoreItemThumbnail AddToWishList() 
        {
            Support.HoverOverWebElement(TestConfiguration.GetDriverInstance(), itemImage);
            if (itemAddToWishListTab.Count == 0)
                throw new NoSuchElementException($"The tab for adding to the wish list for item '{ GetName() }' was not found. " +
                    "Make sure this object is being displayed in the Catalog page");

            itemAddToWishListTab[0].Click();

            var msg = TestConfiguration.GetDriverInstance().FindElements(ERROR_MESSAGE_BOX_LOCATOR);
            if(msg.Count == 0)
                throw new NoSuchElementException("The expected pop up message did not appear");
            if (!msg[0].Text.Equals(ADDED_TO_WISH_LIST))
                throw new ErrorMessageException(msg[0].Text);
            msg[0].FindElement(By.XPath("../../../a[@title='Close']")).Click(); //For closing the popup message that should always appear
            return this;
        }

        public bool IsAddedToWishList()
        {
            Support.HoverOverWebElement(TestConfiguration.GetDriverInstance(), itemImage);
            if (itemAddToWishListTab.Count == 0)
                throw new NoSuchElementException($"The tab for adding to the wish list for item '{ GetName()}' was not found. " +
                        "Make sure this object is being displayed in the Catalog page");
            string attr = itemAddToWishListTab[0].GetAttribute("class");
            return attr.EndsWith("checked");
        }

        public StoreItemThumbnail AddToCompare(bool addToCompare)
        {
            Support.HoverOverWebElement(TestConfiguration.GetDriverInstance(), itemImage);
            if (addToCompare && !IsAddedToCompare())
                itemAddToCompareTab.Click();
            if (!addToCompare && IsAddedToCompare())
                itemAddToCompareTab.Click();
            return this;
        }

        public bool IsAddedToCompare()
        {
            Support.HoverOverWebElement(TestConfiguration.GetDriverInstance(), itemImage);
            var found = itemAddToCompareTab ?? throw new NoSuchElementException($"The tab for adding for comparison for item '{ GetName() }' was not found. " +
                        "Make sure this object is being displayed in the Catalog page");
            string classAttribute = found.GetAttribute("class");
            return classAttribute.EndsWith("checked");
        }

    }

}
